### PR TITLE
retroshare: 0.6.2 -> 0.6.4

### PR DESCRIPTION
**Pull request by Reilithion**
Fixed build with two patches. One fixes the libupnp update. The other
adds some missing QT include lines.

###### Motivation for this change
retroshare package was broken, and out of date. This fixes both.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Infinisil** [comment]
Are these patches self-written or do they have an original source? If the latter (which would be preferable), you should use `fetchpatch` to get them instead

**Comment by Reilithion** [comment]
I had to make these patches myself. At some point, I'd like to send them upstream, but I need to figure out how RetroShare handles library version stuff first. So it comes down to starting a dialogue with that community.

File: pkgs/applications/networking/p2p/retroshare/default.nix [hunk 1]
```diff
@@ -4,22 +4,21 @@
...
   };
 
   # NIX_CFLAGS_COMPILE = [ "-I${glib.dev}/include/glib-2.0" "-I${glib.dev}/lib/glib-2.0/include" "-I${libxml2.dev}/include/libxml2" "-I${sqlcipher}/include/sqlcipher" ];
 
-  patchPhase = ''
-    # Fix build error
-    sed -i 's/UpnpString_get_String(es_event->PublisherUrl)/es_event->PublisherUrl/' \
-      libretroshare/src/upnp/UPnPBase.cpp
-  '';
+  patches = [ ./0000-libupnp-fixes.patch
```
Comment (ryantm) [inline pkgs/applications/networking/p2p/retroshare/default.nix]: Please add comments linking to upstream tracking of these patches, or explanations for why nix-specific patches are needed.

**Review by ryantm — CHANGES_REQUESTED** [review]

**Comment by aanderse** [comment]
@Reilithion are you still interested in proceeding with this PR?

**Comment by Reilithion** [comment]
There has been a commit to the RetroShare repo that may obsolete this pull request. It'd probably be a good idea to check on that. Relevant link: https://github.com/RetroShare/RetroShare/issues/1426

**Comment by chkno** [comment]
So upstream is fixed, but the fixes have not yet made it into a release.  Also, this PR started as a version bump to 0.6.4, but 0.6.5 was released in the mean time (before the libupnp fixes).

Options from here:
1. Leave retroshare broken until 0.6.6 is released.  :(
2. Use these patches against 0.6.4 to get 0.6.4 instead of 0.6.5.  :(
3. Pick a retroshare git revision between releases, like 16115aaf4ae0 (when the libupnp fixes were merged) or whatever's the tip of master today.  These will have had less testing than proper released versions, which may frustrate other users on the network if the version we pick sends them malformed or unexpected data.  :(
4. Rebase upstream's libupnp fix patches to 0.6.5 and duplicate them here until 0.6.6 is released.
5. Rebase upstream's libupnp fix patches to 0.6.5, ask upstream to accept them on a 0.6.5 release branch, and reference that branch here.  Upstream has a bunch of other old branches, so maybe that's normal for them.

**Comment by aanderse** [comment]
ping @chkno @Reilithion - any decision?

**Comment by Reilithion** [comment]
I do not currently have a way to test whether any particular build of Retroshare actually works on NixOS. I therefore can't meaningfully contribute to any option other than 1 or 2. I agree that one of the rebasing options sounds closer to optimal, but I can't make them happen at the moment.

**Comment by chkno** [comment]
I suppose another option would be to 
6. Split libupnp, leaving pkgs/development/libraries/pupnp/default.nix to march forward normally but add pkgs/development/libraries/pupnp/pupnp-1.6.nix that's locked to the 1.6.x branch so retroshare can still use it until retroshare 0.6.6 is released.  It looks like the pupnp 1.6 -> 1.8 version bump was initially reverted over compatibility concerns in 0a92fb93155f0b9fa5e95157dc0b8e96a0c4d118 before later being auto-updated again in ce0f4a1b9383667ad6ac6c2b61a435e0bf650892.

Any thoughts from retroshare maintainer @domenkozar or pupnp compatibility champion @Mic92 ?

**Comment by Mic92** [comment]
There is 0.6.5 out now.

**Comment by chkno** [comment]
0.6.5 has the same [trouble with recent libupnp](https://github.com/RetroShare/RetroShare/issues/1498).

There's an [upstream issue requesting a new release](https://github.com/RetroShare/RetroShare/issues/1808).

**Comment by ehmry** [comment]
0.6.6 is out.

**Comment by Stekke** [comment]
PR for 0.6.6  #124177

**Pull request closed without merging**